v-next/hardhat/src/internal/builtin-plugins/network-manager/type-validation.ts

```diff
@@ -221,54 +220,102 @@ const baseNetworkUserConfigSchema = z.discriminatedUnion("type", [
+      chainType = GENERIC_CHAIN_TYPE,
+      hardfork,
+      chains,
       minGasPrice,
       initialBaseFeePerGas,
       enableTransientStorage,
     } = networkConfig;
 
-    if (hardforkGte(hardfork, HardforkName.LONDON)) {
-      if (minGasPrice !== undefined) {
+    if (hardfork !== undefined && !isValidHardforkName(hardfork, chainType)) {
+      ctx.addIssue({
+        code: z.ZodIssueCode.custom,
+        path: ["hardfork"],
+        message: `Invalid hardfork name ${hardfork} for chainType ${chainType}. Expected ${getHardforks(
+          chainType,
+        ).join(" | ")}.`,
+      });
+    }
+
+    if (chains !== undefined) {
+      let chainIdx = 0;
+      for (const [chainId, chainConfig] of chains) {
+        if (chainConfig.hardforkHistory === undefined) {
+          chainIdx++;
+          continue;
+        }
+
+        const chainLevelChainType = chainConfig.chainType ?? GENERIC_CHAIN_TYPE;
+        let hardforkHistoryIdx = 0;
+        for (const [hardforkName] of chainConfig.hardforkHistory) {
+          if (!isValidHardforkName(hardforkName, chainLevelChainType)) {
+            ctx.addIssue({
+              code: z.ZodIssueCode.custom,
+              path: [
+                "chains",
+                chainIdx,
+                "value",
+                "hardforkHistory",
+                hardforkHistoryIdx,
+                "value",
+              ],
+              message: `Invalid hardfork name ${hardforkName} found in chain ${chainId}. Expected ${getHardforks(
+                chainLevelChainType,
+              ).join(" | ")}.`,
+            });
+          }
+          hardforkHistoryIdx++;
+        }
+        chainIdx++;
+      }
+    }
+
+    const resolvedHardfork = hardfork ?? getLatestHardfork(chainType);
+    if (chainType === L1_CHAIN_TYPE || chainType === GENERIC_CHAIN_TYPE) {
+      if (hardforkGte(resolvedHardfork, L1HardforkName.LONDON, chainType)) {
+        if (minGasPrice !== undefined) {
+          ctx.addIssue({
+            code: z.ZodIssueCode.custom,
+            message:
+              "minGasPrice is not valid for networks with EIP-1559. Try an older hardfork or remove it.",
+          });
+        }
+      } else {
+        if (initialBaseFeePerGas !== undefined) {
+          ctx.addIssue({
+            code: z.ZodIssueCode.custom,
+            message:
+              "initialBaseFeePerGas is only valid for networks with EIP-1559. Try a newer hardfork or remove it.",
+          });
+        }
+      }
+
+      if (
+        !hardforkGte(resolvedHardfork, L1HardforkName.CANCUN, chainType) &&
+        enableTransientStorage === true
+      ) {
         ctx.addIssue({
           code: z.ZodIssueCode.custom,
-          message:
-            "minGasPrice is not valid for networks with EIP-1559. Try an older hardfork or remove it.",
+          message: `'enableTransientStorage' is not supported for hardforks before 'cancun'. Please use a hardfork from 'cancun' onwards to enable this feature.`,
```

Comment:
Is the `enableTransientStorage` config option not obsolete? I thought that was merely added for people to test transient storage before support was added in Cancun

Reply:
Yes, but removing it should not be part of this PR.